### PR TITLE
HTTP transition for asgi

### PR DESCRIPTION
**Pull request by lzchen**
Part of https://github.com/open-telemetry/opentelemetry-python-contrib/issues/2453

Also removing `"net.host.name"` and `"net.host.port"` afrom wsgi based apps for [client.duration](https://github.com/open-telemetry/semantic-conventions/blob/v1.26.0/docs/http/http-metrics.md#metric-httpclientrequestduration) metrics as it was erroneously added in a previous PR.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -25,6 +25,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - `opentelemetry-instrumentation-asgi`, `opentelemetry-instrumentation-fastapi`, `opentelemetry-instrumentation-starlette` Use `tracer` and `meter` of originating components instead of one from `asgi` middleware
   ([#2580](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2580))
+- Populate `{method}` as `HTTP` on `_OTHER` methods from scope
+  ([#2425](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2425))
```
Comment (emdneto) [inline CHANGELOG.md:29]:
```suggestion
  ([#2610](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2610))
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -25,6 +25,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+
+### Added
+
+- `opentelemetry-instrumentation-asgi` Implement new semantic convention opt-in with stable http semantic conventions
+  ([#2425](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2425))
```
Comment (emdneto) [inline CHANGELOG.md:34]:
```suggestion
  ([#2610](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2610))
```

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py [hunk 3]
```diff
@@ -304,31 +336,53 @@ def collect_request_attributes(scope):
         if isinstance(query_string, bytes):
             query_string = query_string.decode("utf8")
         http_url += "?" + urllib.parse.unquote(query_string)
+    result = {}
+
+    scheme = scope.get("scheme")
+    if scheme:
+        _set_http_scheme(result, scheme, sem_conv_opt_in_mode)
+    if server_host:
+        _set_http_host(result, server_host, sem_conv_opt_in_mode)
```
Comment (emdneto) [inline instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:345]: Do we also have to set `_set_http_net_host(result, host, sem_conv_opt_in_mode)` here?

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py [hunk 4]
```diff
@@ -494,30 +561,51 @@ def __init__(
...
+        self.duration_histogram_old = None
+        if _report_old(sem_conv_opt_in_mode):
+            self.duration_histogram_old = self.meter.create_histogram(
+                name=MetricInstruments.HTTP_SERVER_DURATION,
+                unit="ms",
+                description="Duration of HTTP server requests.",
+            )
+        self.duration_histogram_new = None
+        if _report_new(sem_conv_opt_in_mode):
+            self.duration_histogram_new = self.meter.create_histogram(
```
Comment (emdneto) [inline instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:578]: can we use `semconv.create_http_server_request_duration` ?

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py [hunk 5]
```diff
@@ -629,34 +721,69 @@ async def __call__(
...
+                        duration, duration_attrs_old
+                    )
+                if self.duration_histogram_new:
+                    self.duration_histogram_new.record(
+                        duration, duration_attrs_new
+                    )
                 self.active_requests_counter.add(
                     -1, active_requests_count_attrs
                 )
+                print(attributes)
```
Comment (emdneto) [inline instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:760]: Remove this?

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py [hunk 6]
```diff
@@ -629,34 +721,69 @@ async def __call__(
...
+                print(active_requests_count_attrs)
```
Comment (emdneto) [inline instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:761]:
```suggestion
```

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py (hunk 3 above)
Comment (lzchen) [reply instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:345]: Does not seem like `net.host.name` was supported previously in asgi instrumentation with old semconv. New semconv map it to `server.address` which we already populate.

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py (hunk 4 above)
Comment (lzchen) [reply instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:578]: I am purposefully not using apis in semconv that are incubating. Could change in the future.

Comment (emdneto) [reply instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:578]: Got it

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py [hunk 7]
```diff
@@ -629,34 +721,67 @@ async def __call__(
...
+                duration_attrs_new = _parse_duration_attrs(
+                    attributes, _HTTPStabilityMode.HTTP
+                )
+                if self.duration_histogram_old:
+                    self.duration_histogram_old.record(
+                        duration, duration_attrs_old
+                    )
+                if self.duration_histogram_new:
+                    self.duration_histogram_new.record(
+                        duration, duration_attrs_new
```
Comment (emdneto) [inline instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:755]: The duration here is in seconds, right? But in L740 we are getting `duration` in ms. Does it make sense to use duration in seconds here?

**Review by emdneto — COMMENTED** [review]
should we update the status in instrumentation/README.md from `experimental` to `migration` for this instrumentation?

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py (hunk 7 above)
Comment (lzchen) [reply instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:755]: Good catch

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py [hunk 8]
```diff
@@ -304,31 +336,53 @@ def collect_request_attributes(scope):
...
+    if server_host:
+        _set_http_host(result, server_host, sem_conv_opt_in_mode)
+    if port:
+        _set_http_net_host_port(result, port, sem_conv_opt_in_mode)
+    flavor = scope.get("http_version")
+    if flavor:
+        _set_http_flavor_version(result, flavor, sem_conv_opt_in_mode)
+    path = scope.get("path")
+    if path:
+        target = path
```
Comment (ocelotl) [inline instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:353]:
```suggestion
```

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py [hunk 9]
```diff
@@ -304,31 +336,53 @@ def collect_request_attributes(scope):
...
+        _set_http_target(
+            result, target, path, query_string, sem_conv_opt_in_mode
```
Comment (ocelotl) [inline instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:355]:
```suggestion
            result, path, path, query_string, sem_conv_opt_in_mode
```

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py [hunk 10]
```diff
@@ -372,24 +426,30 @@ def get_host_port_url_tuple(scope):
...
+def set_status_code(
+    span,
+    status_code,
+    metric_attributes=None,
+    sem_conv_opt_in_mode=_HTTPStabilityMode.DEFAULT,
+):
     """Adds HTTP response attributes to span using the status_code argument."""
     if not span.is_recording():
         return
+    status_code_str = repr(status_code)
```
Comment (ocelotl) [inline instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:438]: Just curious, why not use `str` here?

**Review by ocelotl — APPROVED** [review]

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py (hunk 10 above)
Comment (lzchen) [reply instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:438]: No real reason. Switched to using `str`.

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/_semconv.py [hunk 11]
```diff
@@ -85,13 +96,12 @@
     SpanAttributes.HTTP_SCHEME,
     SpanAttributes.HTTP_FLAVOR,
     SpanAttributes.HTTP_SERVER_NAME,
-    SpanAttributes.NET_HOST_NAME,
```
Comment (emdneto) [inline opentelemetry-instrumentation/src/opentelemetry/instrumentation/_semconv.py:88]: Can we have a changelog saying we are removing this? I understand it was something wrong introduced in the previous PR

**Review by emdneto — APPROVED** [review]

**Pull request merged**